File: test/auto_ptr.py
'''
>>> from auto_ptr_ext import *
>>> x = X(42)
>>> x.value()
42

>>> look(x), look(x)
(42, 42)

>>> maybe_steal(x, 0)
42
>>> look(x)
42

>>> maybe_steal(x, 1)
42
>>> (not '--broken-auto-ptr' in sys.argv) and look(x) or -1
-1

>>> x = X(69)
>>> steal(x)
69
>>> (not '--broken-auto-ptr' in sys.argv) and look(x) or -1
-1

>>> if not '--broken-auto-ptr' in sys.argv:
...     try: x.value()
...     except TypeError: pass
...     else: print 'expected a TypeError exception'

>>> x = make()
>>> look(x)
77

>>> y = callback(lambda y: y)
>>> y.value()
77

>>> extract(x).value()
77
'''

def run(args = None):
    import sys
    import doctest

    if args is not None:
        sys.argv = args
    return doctest.testmod(sys.modules.get(__name__))
    
if __name__ == '__main__':
    print "running..."
    import sys
    sys.exit(run()[0])

File: test/auto_ptr.cpp
// Copyright David Abrahams 2002. Permission to copy, use,
// modify, sell and distribute this software is granted provided this
// copyright notice appears in all copies. This software is provided
// "as is" without express or implied warranty, and with no claim as
// to its suitability for any purpose.

#include <boost/python/module.hpp>
#include <boost/python/class.hpp>
#include <boost/python/extract.hpp>
#include <boost/python/def.hpp>
#include "test_class.hpp"

#include <memory>

using namespace boost::python;

typedef test_class<> X;

int look(std::auto_ptr<X> const& x)
{
    return (x.get()) ? x->value() : -1;
}

int steal(std::auto_ptr<X> x)
{
    return x->value();
}

int maybe_steal(std::auto_ptr<X>& x, bool doit)
{
    int n = x->value();
    if (doit)
        x.release();
    return n;
}

std::auto_ptr<X> make()
{
    return std::auto_ptr<X>(new X(77));
}

std::auto_ptr<X> callback(object f)
{
    std::auto_ptr<X> x(new X(77));
    return call<std::auto_ptr<X> >(f.ptr(), x);
}

std::auto_ptr<X> extract_(object o)
{
    return extract<std::auto_ptr<X>&>(o);
}

BOOST_PYTHON_MODULE(auto_ptr_ext)
{
    class_<X, std::auto_ptr<X>, boost::noncopyable>("X", init<int>())
        .def("value", &X::value)
        ;
    
    def("look", look);
    def("steal", steal);
    def("maybe_steal", maybe_steal);
    def("make", make);
    def("callback", callback);
    def("extract", extract_);
}

#include "module_tail.cpp"

